Some improvements to the Young environments

giordano opened the pull request:
Saving the progress of my changes so far.

giordano committed:
[multiarch_intel] Do not include `multiarch`

giordano committed:
[gpu_first] openmpi: set scheduler to slurm, add `+legacylaunchers`, remove extra fabrics

giordano committed:
Move `py-alphafold` to gpu-on-gpu

Bazel seems to target the current CPU ISA, sidestepping Spack settings, so we
really need to compile Alphafold on the target machine.

giordano committed:
[gpu-on-gpu] Explicitly request `openblas` as blas provider

For some reason I was getting the resolver decided to link `dbcsr` and
`py-alphafold` to `amdblis` as blas provider:

```
 -   tezgfzo  dbcsr@2.6.0%gcc@12.3.0+cuda~cuda_arch_35_k20x+examples~g2g~ipo+mpi~mpi_f08~opencl~openmp~rocm+shared build_system=cmake build_type=Release cuda_arch=60 generator=ninja smm=libxsmm arch=linux-rhel9-x86_64_v3
 -   ovweai7      ^amdblis@5.0%gcc@12.3.0~aocl_gemm+blas+cblas~ilp64~logging+suphandling~tracing build_system=makefile libs=shared,static threads=none arch=linux-rhel9-x86_64_v3
[+]  vpfhgcm          ^gmake@4.4.1%gcc@12.3.0~guile build_system=generic arch=linux-rhel9-x86_64_v3
 -   ccny7ib      ^amdlibflame@5.0%gcc@12.3.0~debug~ilp64~ipo+lapack2flame~logging+shared+static~tracing build_system=cmake build_type=Release generator=make threads=none vectorization=auto arch=linux-rhel9-x86_64_v3
 -   65o3fuy          ^aocl-utils@5.0%gcc@12.3.0~doc~examples~ipo+shared~tests build_system=cmake build_type=Release generator=make arch=linux-rhel9-x86_64_v3
```

With the explicitly dependency we restrict the options for the resolver and
prevent it from going off road.

giordano committed:
Don't use system `openssl` as external

This casused problems during Rust bootstrap, because Spack-built `pkg-config`
can't find system .pc files. We still need to keep system `openssh` though.

Also, if we don't use `openssl` as external we don't need `curl` either.

giordano committed:
Include second_compiler in `avx2_buildtools`

This ensures the `gcc@12.3.0` spec in the new env reuses the spec from the other
environment.

giordano committed:
Change paths of concrete environments to use the 2025-09 site name

giordano committed:
Include only `multiarch_intel` in `avx512_intel`

giordano committed:
Add ompi412 environment for Young+Michael

giordano committed:
Split avx2 environment into GCC- and Intel-specific environments

giordano committed:
Include more environments in youngmichael

giordano commented on spacksites/spack-env-templates/slurm/youngmichael.yaml:
```diff
@@ -5,9 +5,11 @@ spack:
-  - /lustre/apps/spack/0.23/deploy/2025-08/spack/var/spack/environments/avx2_buildtools
-  - /lustre/apps/spack/0.23/deploy/2025-08/spack/var/spack/environments/avx2
-  - /lustre/apps/spack/0.23/deploy/2025-08/spack/var/spack/environments/avx512
+  - /lustre/apps/spack/0.23/deploy/2025-09/spack/var/spack/environments/avx2_buildtools
+  - /lustre/apps/spack/0.23/deploy/2025-09/spack/var/spack/environments/avx2
+  - /lustre/apps/spack/0.23/deploy/2025-09/spack/var/spack/environments/avx2_intel
+  - /lustre/apps/spack/0.23/deploy/2025-09/spack/var/spack/environments/avx512
+  - /lustre/apps/spack/0.23/deploy/2025-09/spack/var/spack/environments/avx512_intel
```
I'm not sure which environments should be here?  *All* of them?

heatherkellyucl replied:
@giordano Yes, all the environments that make up the youngmichael stack to be used. (So not the compiler versions that we don't want to be used, for example, like first_compiler). This is the part that creates all the module files in the correct place with the correct names, and shouldn't need to add any additional software.

giordano replied:
Done!  Environment created, but there's nothing to do for `install`.  I can regenerate the modules with
```
spack -e youngmichael module tcl refresh --delete-tree
```
if necessary.

giordano committed:
More explicitly require `x86_64_v4` target for root specs in avx512 environments

Similarly to f10f4583051ebe5c977fe974b577e3f8a7734fb3 (in #68),
`packages:all:target` isn't strict enough for forcing the packages to be built
for that target.  At the same time we can't use `packages:all:require` here
because we want to include the `avx2_buildtools` and `multiarch` environments,
which would be incompatible with the requirement of having *all* packages
targeting `x86_64_v4`.  Our solution is to append `target=x86_64_v4` to force
that architecture only for the root specs in the avx512 environments, but allow
dependencies to be reused from the `x86_64_v3` environments.

giordano committed:
Resolve ambiguity in module name generation

Spack uses a "first matching" strategy when choosing the name of the module, in
the collective `youngmichael` environment we have both `openmpi +cuda` and
`openmpi ~cuda` specs, but they had conflicting module names because only the
`openmpi` projection was matched, not also the `^cuda` one.  By adding an
explicit `openmpi +cuda` projection before `openmpi` we can resolve the
ambiguity.

giordano commented:
This should be ready for review.  For the record, these are more or less the commands I ran:
```shell
source spacksites/myriad-utilities/init-spacksites-rhel9.sh
site_name=2025-09
spacksites/spacksites --config-file spacksites/settings/deploy_sites.ini create ${site_name}
# build first compiler
spacksites/spacksites --config-file spacksites/settings/deploy_sites.ini install-env ${site_name} compiler first_compiler.yaml
eval $(spacksites/spacksites --config-file spacksites/settings/deploy_sites.ini spack-setup-env ${site_name})

# second_compiler gcc@12.3.0%gcc@12.3.0
spack env create second_compiler spacksites/spack-env-templates/slurm/second_compiler.yaml
spack -e second_compiler concretize -U -f
spack -e second_compiler install --only-concrete
spack -e second_compiler buildcache push /lustre/apps/spack/0.23/buildcache

spack compiler add $(spack -e second_compiler location -i gcc@12.3.0%gcc@12.3.0)

# check list of compilers
spack config blame compilers
# if there are multiple gcc@12.3.0, and in particular if there's a gcc@12.3.0 compiled by gcc@12.3.0 in `site/linux`, and then another one compiled by gcc@11.5.0 (the "system" one) in `site`, remove the latter
spack compiler remove --scope site gcc@12.3.0
# Double check that now the list of compilers includes only the original gcc@11.5.0 and the second gcc@12.3.0
spack config blame compilers

# avx2_buildtools
spack env create avx2_buildtools spacksites/spack-env-templates/slurm/avx2_buildtools.yaml
spack -e avx2_buildtools concretize -U -f
spack -e avx2_buildtools env depfile -o Makefile
make --output-sync=target -j40

# add the newly built intel compiler to the list of compilers in site/linux
spack compiler add $(spack -e avx2_buildtools location -i intel-oneapi-compilers)/compiler/latest/bin

# multiarch
spack env create multiarch spacksites/spack-env-templates/slurm/multiarch.yaml
spack -e multiarch concretize -U -f
spack -e multiarch env depfile -o Makefile
make --output-sync=target -j40

# intel multiarch
spack env create multiarch_intel spacksites/spack-env-templates/slurm/multiarch_intel.yaml
spack -e multiarch_intel concretize -U -f
spack -e multiarch_intel env depfile -o Makefile
make --output-sync=target -j40

# avx512
spack env create avx512 spacksites/spack-env-templates/slurm/avx512.yaml
spack -e avx512 concretize -U -f
spack -e avx512 env depfile -o Makefile
make --output-sync=target -j40

# avx512_intel
spack env create avx512_intel spacksites/spack-env-templates/slurm/avx512_intel.yaml
spack -e avx512_intel concretize -U -f
spack -e avx512_intel env depfile -o Makefile
make --output-sync=target -j40

# gpu_first
spack env create gpu_first spacksites/spack-env-templates/slurm/gpu_first.yaml
spack -e gpu_first concretize -U -f
spack -e gpu_first env depfile -o Makefile
make --output-sync=target -j40

# gpu-on-gpu
spack env create gpu-on-gpu spacksites/spack-env-templates/slurm/gpu-on-gpu.yaml
spack -e gpu-on-gpu concretize -U -f
spack -e gpu-on-gpu env depfile -o Makefile
make --output-sync=target -j40

# gpu_third
spack env create gpu_third spacksites/spack-env-templates/slurm/gpu_third.yaml
spack -e gpu_third concretize -U -f
spack -e gpu_third env depfile -o Makefile
make --output-sync=target -j40

# ompi412-youngmichael
spack env create ompi412-youngmichael spacksites/spack-env-templates/slurm/ompi412-youngmichael.yaml
spack -e ompi412-youngmichael concretize -U -f
spack -e ompi412-youngmichael env depfile -o Makefile
make --output-sync=target -j40

# avx2
spack env create avx2 spacksites/spack-env-templates/slurm/avx2.yaml
spack -e avx2 concretize -U -f
spack -e avx2 env depfile -o Makefile
make --output-sync=target -j40

# avx2_intel
spack env create avx2_intel spacksites/spack-env-templates/slurm/avx2_intel.yaml
spack -e avx2_intel concretize -U -f
spack -e avx2_intel env depfile -o Makefile
make --output-sync=target -j40

# youngmichael: nothing to install, but regenerate all the modules
spack env create youngmichael spacksites/spack-env-templates/slurm/youngmichael.yaml
spack -e youngmichael concretize -U -f
spack -e youngmichael module tcl refresh --delete-tree
```

heatherkellyucl commented:
Looking at the modules we end up with, there are still some things that appear to be issues:

<img width="932" height="466" alt="Screenshot 2025-10-01 at 17 33 10" src="https://github.com/user-attachments/assets/229f4afc-f95b-4048-b7c3-989ae2625493" />

We have a bunch of top-level apps for `x86_64_v4` but not for `x86_64_v3`. We have some versions of gromacs etc for both but then some not. We have two namds for v4 but one for v3. We have no lammps for v3.

We don't have any orca 6.0.1. 

Then looking at MPI, something in the v3 and v4 stack has built its own openmpi 4.1.7 which we really don't want because it almost certainly has the wrong options. It is a hidden module with a hash because it isn't a root spec.

```
[cceahke@login21 verbs]$ module avail mpi
--------------------------------------------------- /apps/spack/0.23/deploy/2025-09/modules/linux-rhel9-x86_64_v4 ---------------------------------------------------
mpi/openmpi/4.1.7/gcc-12.3.0-b7b7kuf  

--------------------------------------------------- /apps/spack/0.23/deploy/2025-09/modules/linux-rhel9-x86_64_v3 ---------------------------------------------------
mpi/intel-oneapi-mpi/2021.14.0/intel-oneapi-2024.2.1  mpi/openmpi/4.1.6-cuda/gcc-12.3.0  mpi/openmpi/4.1.6/oneapi-2024.2.1     
mpi/openmpi/4.1.2/gcc-12.3.0                          mpi/openmpi/4.1.6/gcc-12.3.0       mpi/openmpi/4.1.7/gcc-12.3.0-koebyjx 
```

Found in:
```
/apps/spack/0.23/deploy/2025-09/modules/linux-rhel9-x86_64_v4/dbcsr/2.7.0-openmpi/gcc-12.3.0-lriirqb:    module load mpi/openmpi/4.1.7/gcc-12.3.0-b7b7kuf
/apps/spack/0.23/deploy/2025-09/modules/linux-rhel9-x86_64_v4/dbcsr/2.7.0-openmpi/gcc-12.3.0-lriirqb:    depends-on mpi/openmpi/4.1.7/gcc-12.3.0-b7b7kuf
/apps/spack/0.23/deploy/2025-09/modules/linux-rhel9-x86_64_v4/mpi/openmpi/4.1.7/.modulerc:module-hide --soft --hidden-loaded mpi/openmpi/4.1.7/gcc-12.3.0-b7b7kuf
/apps/spack/0.23/deploy/2025-09/modules/linux-rhel9-x86_64_v4/hdf5/1.14.3-openmpi/gcc-12.3.0:    module load mpi/openmpi/4.1.7/gcc-12.3.0-b7b7kuf
/apps/spack/0.23/deploy/2025-09/modules/linux-rhel9-x86_64_v4/hdf5/1.14.3-openmpi/gcc-12.3.0:    depends-on mpi/openmpi/4.1.7/gcc-12.3.0-b7b7kuf
/apps/spack/0.23/deploy/2025-09/modules/linux-rhel9-x86_64_v4/netlib-scalapack/2.2.0-openmpi/gcc-12.3.0-eydcuqn:    module load mpi/openmpi/4.1.7/gcc-12.3.0-b7b7kuf
/apps/spack/0.23/deploy/2025-09/modules/linux-rhel9-x86_64_v4/netlib-scalapack/2.2.0-openmpi/gcc-12.3.0-eydcuqn:    depends-on mpi/openmpi/4.1.7/gcc-12.3.0-b7b7kuf
/apps/spack/0.23/deploy/2025-09/modules/linux-rhel9-x86_64_v4/cp2k/2024.3-openmpi/gcc-12.3.0:    module load mpi/openmpi/4.1.7/gcc-12.3.0-b7b7kuf
/apps/spack/0.23/deploy/2025-09/modules/linux-rhel9-x86_64_v4/cp2k/2024.3-openmpi/gcc-12.3.0:    depends-on mpi/openmpi/4.1.7/gcc-12.3.0-b7b7kuf
/apps/spack/0.23/deploy/2025-09/modules/linux-rhel9-x86_64_v4/gromacs/2023.5-openmpi-plumed/gcc-12.3.0:    module load mpi/openmpi/4.1.7/gcc-12.3.0-b7b7kuf
/apps/spack/0.23/deploy/2025-09/modules/linux-rhel9-x86_64_v4/gromacs/2023.5-openmpi-plumed/gcc-12.3.0:    depends-on mpi/openmpi/4.1.7/gcc-12.3.0-b7b7kuf
/apps/spack/0.23/deploy/2025-09/modules/linux-rhel9-x86_64_v4/plumed/2.9.2-openmpi-cuda/gcc-12.3.0-2b2c5bt:    module load mpi/openmpi/4.1.7/gcc-12.3.0-b7b7kuf
/apps/spack/0.23/deploy/2025-09/modules/linux-rhel9-x86_64_v4/plumed/2.9.2-openmpi-cuda/gcc-12.3.0-2b2c5bt:    depends-on mpi/openmpi/4.1.7/gcc-12.3.0-b7b7kuf
/apps/spack/0.23/deploy/2025-09/modules/linux-rhel9-x86_64_v3/mpi/openmpi/4.1.7/.modulerc:module-hide --soft --hidden-loaded mpi/openmpi/4.1.7/gcc-12.3.0-koebyjx
/apps/spack/0.23/deploy/2025-09/modules/linux-rhel9-x86_64_v3/hdf5/1.14.3-openmpi/gcc-12.3.0:    module load mpi/openmpi/4.1.7/gcc-12.3.0-koebyjx
/apps/spack/0.23/deploy/2025-09/modules/linux-rhel9-x86_64_v3/hdf5/1.14.3-openmpi/gcc-12.3.0:    depends-on mpi/openmpi/4.1.7/gcc-12.3.0-koebyjx
/apps/spack/0.23/deploy/2025-09/modules/linux-rhel9-x86_64_v3/netlib-scalapack/2.2.0-openmpi/gcc-12.3.0-5em4lbr:    module load mpi/openmpi/4.1.7/gcc-12.3.0-koebyjx
/apps/spack/0.23/deploy/2025-09/modules/linux-rhel9-x86_64_v3/netlib-scalapack/2.2.0-openmpi/gcc-12.3.0-5em4lbr:    depends-on mpi/openmpi/4.1.7/gcc-12.3.0-koebyjx
```

We also need a separate first_compiler.yaml and second_compiler.yaml for this stack, because we need those for the other stacks too, but without the target arch addition. So those should be put into `first_compiler_youngmichael.yaml` and `second_compiler_youngmichael.yaml` for this stack only.

heatherkellyucl commented:
I had a discussion earlier this morning that we need to work out what module namings make sense for this stack for users and whether the v3 should have `/avx2` appended or something if doable, since if you `module use` the modules like this so they're in sections, then they don't have the long `linux-rhel9-x86_64_v3` prepended to all the names, but they do have name clashes. That's not such an immediate issue though.

giordano commented:
> We have no lammps for v3.

That was weird, because lammps _is_ in avx2 https://github.com/UCL-ARC/hpc-spack/blob/2243e8fa34284b4af31269cc6aed7a146e19961d/spacksites/spack-env-templates/slurm/avx2.yaml#L29 and the lock file at `/apps/spack/0.23/deploy/2025-09/spack/var/spack/environments/avx2/spack.lock` shows it.  But I just concretized the avx2 environment and noticed a bunch of packages, including lammps, are not installed.  I'm not really sure what happened here, if I missed to reinstall the full environment (I was fairly sure I did...) or what, in any case I'm doing the install again now.

> Then looking at MPI, something in the v3 and v4 stack has built its own openmpi 4.1.7 which we really don't want because it almost certainly has the wrong options. It is a hidden module with a hash because it isn't a root spec.

Uhm, if a different openmpi was installed it may be because some other packages have compatibility requirements which don't match the build we have, but if we want to enforce certain options everywhere that has to be done with [package requirements](https://spack.readthedocs.io/en/latest/packages_yaml.html#package-requirements).

giordano commented:
>  I'm not really sure what happened here, if I missed to reinstall the full environment (I was fairly sure I did...) or what, in any case I'm doing the install again now.

Ah, I think I understand what may have happened, because it just happened now: installation of orca fails because it can't find the source, but I missed that that operation failed because it was very high up in the log, other packages were installed successfully after orca, but not _all_ of them.

giordano commented:
Also, one charmpp is unahppy:
```console
$ cat /lustre/apps/spack/0.23/deploy/2025-09/build_stage/ccspapp/spack-stage-charmpp-6.10.2-ei7a35gzqbkibqoxmzbyxjvwb3chzbq6/spack-build-out.txt
==> charmpp: Executing phase: 'install'
==> [2025-10-01-22:03:10.959432] './build' 'LIBS' 'ucx-linux-x86_64' 'gcc' 'gfortran' '-j6' '--destination=/lustre/apps/spack/0.23/deploy/2025-09/spack/opt/spack/linux-rhel9-x86_64_v3/gcc-12.3.0/charmpp-6.10.2-ei7a35gzqbkibqoxmzbyxjvwb3chzbq6' 'openpmix' '--basedir=/usr' '--basedir=/usr' 'smp' '--build-shared' '--with-production'
Error> option: openpmix is not supported in this version!
Supported compilers: clang craycc gcc icc iccstatic msvc pgcc xlc xlc64
Supported options: bigemulator bigsim common cuda flang gfortran ifort nolb omp ompipmix ooc papi pedantic perftools persistent pgf90 slurmpmi slurmpmi2 smp tsan
```

heatherkellyucl commented:
That's the namd 2.14 which ought to have been removed entirely.

heatherkellyucl commented:
Was wondering if the openmpi might be because cp2k _doesn't_ want a cuda-aware MPI but haven't looked more closely.

heatherkellyucl commented:
Going to merge this now and then go and move things around in the main branch - I want separate yamls for the `2025-05` deployed environments vs the `2025-09_youngmichael` environments so the second don't affect the first.
